content/docs/futures/combinators.md

```diff
@@ -0,0 +1,202 @@
+---
+title: "Combinators"
+weight : 2040
+menu:
+  docs:
+    parent: futures
+---
+
+Often times, Future implementations follow similar patterns. To help reduce
+boilerplate, the `futures` crate provides a number of utilities, called
+"combinators", that abstract these patterns. Many of these combinators exist as
+functions on the [`Future`] trait.
+
+# Building blocks
+
+Let's revisit the future implementations from the previous pages and see how
+they can be simplified by using combinators.
+
+## `map`
+
+The [`map`] combinator takes a future and returns a new future that applies a
+function to the value yielded by the first future.
+
+This was the `Display` future [`previously`][display-fut] implemented:
+
+```rust
+# #![deny(deprecated)]
+# #[macro_use]
+# extern crate futures;
+# extern crate tokio;
+#
+# use futures::{Future, Async, Poll};
+# use std::fmt;
+#
+# struct Display<T>(T);
+#
+impl<T> Future for Display<T>
+where
+    T: Future,
+    T::Item: fmt::Display,
+{
+    type Item = ();
+    type Error = T::Error;
+
+    fn poll(&mut self) -> Poll<(), T::Error> {
+        let value = try_ready!(self.0.poll());
+        println!("{}", value);
+        Ok(Async::Ready(()))
+    }
+}
+
+fn main() {
+# let HelloWorld = futures::future::ok::<_, ()>("hello");
+    let future = Display(HelloWorld);
+    tokio::run(future);
+}
+```
+
+With the `map` combinator, it becomes:
+
+```rust
+# #![deny(deprecated)]
+extern crate tokio;
+extern crate futures;
+
+use futures::Future;
+
+fn main() {
+# let HelloWorld = futures::future::ok::<_, ()>("hello");
+    let future = HelloWorld.map(|value| {
+        println!("{}", value);
+    });
+
+    tokio::run(future);
+}
+```
+
+This is how `map` is implemented:
+
+```rust
+# #![deny(deprecated)]
+# #[macro_use]
+# extern crate futures;
+# use futures::{Future, Async, Poll};
+pub struct Map<A, F> where A: Future {
+    future: A,
+    f: Option<F>,
+}
+
+impl<U, A, F> Future for Map<A, F>
+    where A: Future,
+          F: FnOnce(A::Item) -> U,
+{
+    type Item = U;
+    type Error = A::Error;
+
+    fn poll(&mut self) -> Poll<U, A::Error> {
+        let value = try_ready!(self.future.poll());
+        let f = self.f.take().expect("cannot poll Map twice");
+
+        Ok(Async::Ready(f(value)))
+    }
+}
+# fn main() {}
+```
+
+Comparing `Map` with our `Display` implementation, it is clear how they both are
+very similar. Where `Display` calls `println!`, `Map` passes the value to the
+function.
+
+## `and_then`
```

Comment:
It might be worth adding a note about the use of `.map(...).and_then(...)`. At some point I got tripped up wondering why the `and_then` wasn't being called and after a while of debugging it realized it was because the underlying stream + map combination wasn't ever finishing